src/check_ast.rs

```diff
@@ -1946,71 +1956,100 @@ impl<'a> Checker<'a> {
                 }
-
-                first_iter = false;
-                in_generator = matches!(scope.kind, ScopeKind::Generator);
-                import_starred = import_starred || scope.import_starred;
             }
+            _ => return,
+        };
 
-            if import_starred {
-                if self.settings.enabled.contains(&CheckCode::F405) {
-                    let mut from_list = vec![];
-                    for scope_index in self.scope_stack.iter().rev() {
-                        let scope = &self.scopes[*scope_index];
-                        for (name, binding) in scope.values.iter() {
-                            if matches!(binding.kind, BindingKind::StarImportation) {
-                                from_list.push(name.to_string());
-                            }
-                        }
-                    }
-                    from_list.sort();
+        let scope_id = self.scopes[*(self.scope_stack.last().expect("No current scope found."))].id;
 
-                    self.add_check(Check::new(
-                        CheckKind::ImportStarUsage(id.clone(), from_list),
-                        Range::from_located(expr),
-                    ));
+        let mut not_found = None;
+        let mut first_iter = true;
+        let mut in_generator = false;
+        let mut import_starred = false;
+        for scope_index in self.scope_stack.iter().rev() {
+            let scope = &mut self.scopes[*scope_index];
+            if matches!(scope.kind, ScopeKind::Class(_)) {
+                if ids.len() == 1 && ids.get(0).map(|e| &e.0) == Some(&"__class__".to_owned()) {
+                    return;
+                } else if !first_iter && !in_generator {
+                    continue;
                 }
-                return;
             }
 
-            if self.settings.enabled.contains(&CheckCode::F821) {
-                // Allow __path__.
-                if self.path.ends_with("__init__.py") && id == "__path__" {
+            // Starting from the most specific access go through all modules and
+            // check if they have been imported. We remember the identifier that
+            // didn't match to accurately create a check later.
+            for (ident, expr) in &ids {
+                if try_mark_used(scope, scope_id, ident, expr) {
                     return;
                 }
 
-                // Avoid flagging if NameError is handled.
-                if let Some(handler_names) = self.except_handlers.last() {
-                    if handler_names.contains(&"NameError".to_string()) {
-                        return;
+                not_found = Some((ident, expr));
+            }
+
+            first_iter = false;
+            in_generator = matches!(scope.kind, ScopeKind::Generator);
+            import_starred = import_starred || scope.import_starred;
+        }
+
+        let (not_found_ident, not_found_expr) = not_found.unwrap();
```

Comment:
Is this safe? What assumptions is it making? (That there's at least one member in `ids`, and at least one member in `self.scope_stack`?)

Reply:
Yes to both.

At least one member in `ids` should be safe to assume. We're either called with a `Name` or `Attribute`, return otherwise. 
You are right about the `scope_stack`. If that can be empty, we would crash here. I'm not too familiar with that yet, can this be empty?

Reply:
I could initialize `not_found` with the first entry of `ids`, which is then either the name or the full expression for an attribute (e.g. `multiprocessing.pool.Pool`).

If we find a match in the imports, we return early.

If we don't, we'll report the whole expression/identifier as unknown.

Reply:
Okay, removed the unwrap and instead pick the first entry from `ids`. If that is empty, we return.